File: Stack/decode_string.cpp
#include <bits/stdc++.h>
using namespace std;

string decodeString(string s) {
    stack<int> numSt;
    stack<string> strSt;
    string curr = "";
    int num = 0;

    for(char c : s) {
        if(isdigit(c)) {
            num = num * 10 + (c - '0');
        } else if(c == '[') {
            numSt.push(num);
            strSt.push(curr);
            num = 0;
            curr = "";
        } else if(c == ']') {
            int k = numSt.top(); numSt.pop();
            string prev = strSt.top(); strSt.pop();
            while(k--) prev += curr;
            curr = prev;
        } else {
            curr += c;
        }
    }
    return curr;
}

File: Stack/DecodeString.java
import java.util.*;

class DecodeString {
    public String decodeString(String s) {
        Stack<Integer> numStack = new Stack<>();
        Stack<String> strStack = new Stack<>();
        String curr = "";
        int num = 0;

        for (char c : s.toCharArray()) {
            if (Character.isDigit(c)) {
                num = num * 10 + (c - '0');
            } else if (c == '[') {
                numStack.push(num);
                strStack.push(curr);
                num = 0;
                curr = "";
            } else if (c == ']') {
                int k = numStack.pop();
                String prev = strStack.pop();
                curr = prev + curr.repeat(k);
            } else {
                curr += c;
            }
        }
        return curr;
    }
}
